File: src/updateFlowCore.ts
import type { CheckResult } from './type';

// 更新流程的纯决策层:无 IO、不依赖 react-native、无模块级状态,所有输入均由
// 参数传入(仅允许 type-only import)。这一层必须保持可在裸 JS 引擎中求值——
// 它就是未来 guardian bundle 的编译单元(NATIVE_CHECKUPDATE_DESIGN §5),
// IO(HTTP/下载/落盘)由调用方(现在是 client.ts,将来是原生编排)执行。

/* eslint-disable no-fallthrough */
/* eslint-disable no-bitwise */
export function murmurhash3_32_gc(key: string, seed = 0) {
  let remainder: number,
    bytes: number,
    h1: number,
    h1b: number,
    c1: number,
    c2: number,
    k1: number,
    i: number;

  remainder = key.length & 3; // key.length % 4
  bytes = key.length - remainder;
  h1 = seed;
  c1 = 0xcc9e2d51;
  c2 = 0x1b873593;
  i = 0;

  while (i < bytes) {
    k1 =
      (key.charCodeAt(i) & 0xff) |
      ((key.charCodeAt(++i) & 0xff) << 8) |
      ((key.charCodeAt(++i) & 0xff) << 16) |
      ((key.charCodeAt(++i) & 0xff) << 24);
    ++i;

    k1 =
      ((k1 & 0xffff) * c1 + ((((k1 >>> 16) * c1) & 0xffff) << 16)) & 0xffffffff;
    k1 = (k1 << 15) | (k1 >>> 17);
    k1 =
      ((k1 & 0xffff) * c2 + ((((k1 >>> 16) * c2) & 0xffff) << 16)) & 0xffffffff;

    h1 ^= k1;
    h1 = (h1 << 13) | (h1 >>> 19);
    h1b =
      ((h1 & 0xffff) * 5 + ((((h1 >>> 16) * 5) & 0xffff) << 16)) & 0xffffffff;
    h1 = (h1b & 0xffff) + 0x6b64 + ((((h1b >>> 16) + 0xe654) & 0xffff) << 16);
  }

  k1 = 0;

  switch (remainder) {
    // biome-ignore lint/suspicious/noFallthroughSwitchClause: MurmurHash fallthrough
    case 3:
      k1 ^= (key.charCodeAt(i + 2) & 0xff) << 16;
    // biome-ignore lint/suspicious/noFallthroughSwitchClause: MurmurHash fallthrough
    case 2:
      k1 ^= (key.charCodeAt(i + 1) & 0xff) << 8;
    case 1:
      k1 ^= key.charCodeAt(i) & 0xff;

      k1 =
        ((k1 & 0xffff) * c1 + ((((k1 >>> 16) * c1) & 0xffff) << 16)) &
        0xffffffff;
      k1 = (k1 << 15) | (k1 >>> 17);
      k1 =
        ((k1 & 0xffff) * c2 + ((((k1 >>> 16) * c2) & 0xffff) << 16)) &
        0xffffffff;
      h1 ^= k1;
  }

  h1 ^= key.length;

  h1 ^= h1 >>> 16;
  h1 =
    ((h1 & 0xffff) * 0x85ebca6b +
      ((((h1 >>> 16) * 0x85ebca6b) & 0xffff) << 16)) &
    0xffffffff;
  h1 ^= h1 >>> 13;
  h1 =
    ((h1 & 0xffff) * 0xc2b2ae35 +
      ((((h1 >>> 16) * 0xc2b2ae35) & 0xffff) << 16)) &
    0xffffffff;
  h1 ^= h1 >>> 16;

  return h1 >>> 0;
}

export function isInRollout(rollout: number, uuid: string) {
  return murmurhash3_32_gc(uuid) % 100 < rollout;
}

export function joinUrls(paths: string[], fileName?: string) {
  if (fileName) {
    return paths.map((path) => {
      const normalizedPath = path.replace(/\/+$/, '');
      // Keep explicit http(s) URLs for local/self-hosted update sources.
      const baseUrl = /^[a-z][a-z0-9+.-]*:\/\//i.test(normalizedPath)
        ? normalizedPath
        : `https://${normalizedPath}`;
      return `${baseUrl}/${fileName}`;
    });
  }
}

export const dedupeEndpoints = (
  endpoints: Array<string | null | undefined>
): string[] => {
  const result: string[] = [];
  const visited = new Set<string>();

  for (const endpoint of endpoints) {
    if (!endpoint || visited.has(endpoint)) {
      continue;
    }
    visited.add(endpoint);
    result.push(endpoint);
  }

  return result;
};

/**
 * The endpoint plan: dedupe, then move the sampled pick to the front (load
 * spreading), keeping the rest in configured order as the fallback sequence.
 * `randomSample` ∈ [0, 1) is injected by the caller — this layer cannot draw
 * randomness itself.
 */
export function orderEndpointCandidates(
  endpoints: Array<string | null | undefined>,
  randomSample = 0
): string[] {
  const deduped = dedupeEndpoints(endpoints);
  if (deduped.length < 2) {
    return deduped;
  }
  const first = Math.max(
    0,
    Math.min(
      Number.isFinite(randomSample)
        ? Math.floor(randomSample * deduped.length)
        : 0,
      deduped.length - 1
    )
  );
  return [
    deduped[first],
    ...deduped.slice(0, first),
    ...deduped.slice(first + 1),
  ];
}

export interface CheckRequestInput {
  /** Effective native package version (overridePackageVersion already applied). */
  packageVersion: string;
  /** Hash of the currently running JS version ('' when on the packaged bundle). */
  currentVersion?: string;
  buildTime: string;
  cInfo: Record<string, string>;
  supportedDiffVersion?: number;
  /** '' or undefined while unknown — the field is then omitted. */
  bundleHash?: string;
  isDev?: boolean;
  extra?: Record<string, any>;
}

export function buildCheckRequestBody({
  packageVersion,
  currentVersion,
  buildTime,
  cInfo,
  supportedDiffVersion,
  bundleHash,
  isDev,
  extra,
}: CheckRequestInput): Record<string, any> {
  // Caller extras are spread FIRST so they can add fields (channel, custom
  // filters, the debug-channel toHash) but never override the SDK identity
  // fields the server keys its decision on: a stray `packageVersion` or
  // `hash` in extra would otherwise make the server hand out a package for
  // the wrong binary.
  const body: Record<string, any> = {
    ...extra,
    packageVersion,
    hash: currentVersion,
    buildTime,
    cInfo,
  };
  // 可消费的 diff 轨道版本(2 = hdiffv2 轨道),服务端据此门控下发
  if (supportedDiffVersion) {
    body.diffV = supportedDiffVersion;
  } else {
    delete body.diffV;
  }
  if (bundleHash) {
    body.bundleHash = bundleHash;
  } else {
    delete body.bundleHash;
  }
  if (isDev) {
    delete body.buildTime;
  }
  // Transitional dual form: the legacy root-level spread above is what
  // servers before 2026-08-30 read; the nested copy is what new servers read
  // first (extra.toHash wins there) and lets caller keys stop sharing the
  // SDK's namespace once the root spread is retired.
  if (extra) {
    // Match JSON.stringify semantics: undefined values do not exist on the
    // wire, so they must not create a nested copy either (the native check
    // cache compares serialized requests).
    const defined: Record<string, any> = {};
    for (const [key, value] of Object.entries(extra)) {
      if (value !== undefined) {
        defined[key] = value;
      }
    }
    if (Object.keys(defined).length > 0) {
      body.extra = defined;
    }
  }
  return body;
}

export interface CheckFingerprintInput {
  appKey?: string;
  endpoints?: string[];
  queryUrls?: string[];
  uuid?: string;
  /** The exact serialized request body (identity fields + extra). */
  body: string;
}

/**
 * Identity of one update check. Two checks may share an in-flight response
 * only when every input that could change the server's answer is identical.
 */
export function buildCheckFingerprint({
  appKey,
  endpoints,
  queryUrls,
  uuid,
  body,
}: CheckFingerprintInput): string {
  return JSON.stringify([
    appKey ?? '',
    endpoints ?? [],
    queryUrls ?? [],
    uuid ?? '',
    body,
  ]);
}

/**
 * Whether a failed download attempt is worth retrying on another mirror of
 * the same artifact. Transport failures (DNS, connect, timeout, truncated
 * body, 5xx) and corrupt bytes are; a patch that downloaded fine but could
 * not be applied is not — every mirror serves the same bytes, so the next
 * strategy is the only way forward.
 */
export function isMirrorRetryableCode(code?: string): boolean {
  return code !== 'PATCH_FAILED';
}

/**
 * Schema gate for a checkUpdate response (mirrors C++ IsValidCheckResult): a
 * JSON object carrying at least one verdict field of the expected type —
 * `upToDate` / `update` / `expired` (boolean) or `paused` (string). A 200
 * with `{"error": ...}`, an array or an HTML page that happened to parse is
 * a failed endpoint, not a verdict, and must not stop the endpoint fallback.
 */
export function isValidCheckResult(value: unknown): value is CheckResult {
  if (!value || typeof value !== 'object' || Array.isArray(value)) {
    return false;
  }
  const root = value as Record<string, unknown>;
  return (
    typeof root.upToDate === 'boolean' ||
    typeof root.update === 'boolean' ||
    typeof root.expired === 'boolean' ||
    typeof root.paused === 'string'
  );
}

export interface UpdateIdentity {
  packageVersion: string;
  currentVersion?: string;
  /** Stable client uuid — the gray-release bucketing key. */
  uuid: string;
}

export function resolveCheckResult(
  rootInfo: CheckResult,
  identity: UpdateIdentity,
  log: (...args: any[]) => void = () => {}
): CheckResult {
  const { expVersion, ...rootResult } = rootInfo;
  const rollout = expVersion?.config?.rollout?.[identity.packageVersion];
  if (rootResult.update && expVersion && typeof rollout === 'number') {
    if (isInRollout(rollout, identity.uuid)) {
      log(`${expVersion.name} in ${rollout}% rollout, continue`);
      if (
        typeof expVersion.hash === 'string' &&
        expVersion.hash.length > 0 &&
        expVersion.hash === identity.currentVersion
      ) {
        return { upToDate: true };
      }
      const info: CheckResult = {
        update: true,
        ...expVersion,
      };
      if (rootResult.paths) {
        info.paths = rootResult.paths;
      }
      return info;
    }
    log(`${expVersion.name} not in ${rollout}% rollout, ignored`);
  }
  if (
    rootResult.update &&
    typeof rootResult.hash === 'string' &&
    rootResult.hash.length > 0 &&
    rootResult.hash === identity.currentVersion
  ) {
    return { upToDate: true };
  }
  return rootResult;
}

/**
 * Whether the native orchestrator should activate a downloaded version for
 * the next launch: either the client opted in via its silent strategies
 * (afterDownload === 'setNeedUpdate'), or the server marked this version
 * `config.forceBoot` — the per-version remote override that closes the
 * brick-rescue gap for alert-strategy apps (a bricked device never runs JS,
 * so activation cannot wait for it). Native-only: the JS side's interactive
 * strategies are not consulted and not affected. The device-local
 * rolledBackVersion guard in decideDownload still wins over forceBoot, and
 * the activated version keeps the first_time crash-protection rollback.
 */
export function shouldActivateAfterDownload(
  info: CheckResult,
  afterDownload?: string
): boolean {
  return afterDownload === 'setNeedUpdate' || !!info?.config?.forceBoot;
}

export type DownloadStrategyType = 'diff' | 'pdiff' | 'full';

export interface DownloadAttempt {
  type: DownloadStrategyType;
  /** Candidate URLs (paths × file name); the executor probes and picks one. */
  urls: string[];
}

export interface DownloadPlan {
  hash: string;
  /** Ordered attempts: incremental first, full as the last resort. */
  attempts: DownloadAttempt[];
  /** Dev-only: nothing to fetch — treat as an immediate no-op success. */
  devNoop: boolean;
}

export type DownloadDecision =
  | {
      action: 'none';
      reason: 'noUpdate' | 'alreadyCurrent' | 'rolledBack' | 'noArtifact';
    }
  | ({ action: 'download' } & DownloadPlan);

export function decideDownload(
  info: CheckResult,
  identity: { currentVersion?: string; rolledBackVersion?: string },
  isDev = false
): DownloadDecision {
  const { hash, diff, pdiff, full, paths = [] } = info;
  if (!info.update || !hash) {
    return { action: 'none', reason: 'noUpdate' };
  }
  if (hash === identity.currentVersion) {
    return { action: 'none', reason: 'alreadyCurrent' };
  }
  if (identity.rolledBackVersion && hash === identity.rolledBackVersion) {
    return { action: 'none', reason: 'rolledBack' };
  }
  const attempts: DownloadAttempt[] = [];
  // Incremental artifacts only exist against release builds; dev goes
  // straight to full (or the no-op below when there is none).
  if (!isDev) {
    const diffUrls = joinUrls(paths, diff);
    if (diffUrls?.length) {
      attempts.push({ type: 'diff', urls: diffUrls });
    }
    const pdiffUrls = joinUrls(paths, pdiff);
    if (pdiffUrls?.length) {
      attempts.push({ type: 'pdiff', urls: pdiffUrls });
    }
  }
  const fullUrls = joinUrls(paths, full);
  if (fullUrls?.length) {
    attempts.push({ type: 'full', urls: fullUrls });
  }
  const devNoop = !!isDev && !fullUrls?.length;
  if (!attempts.length && !devNoop) {
    return { action: 'none', reason: 'noArtifact' };
  }
  return {
    action: 'download',
    hash,
    attempts,
    devNoop,
  };
}

File: cpp/update_flow_core/update_flow_core.cpp
#include "update_flow_core.h"

#include <cmath>

namespace updateflow {

using flowjson::Value;

namespace {

// The low byte of every UTF-16 code unit of `utf8`, which is exactly what the
// TS reference hashes (`key.charCodeAt(i) & 0xff` over `key.length` units).
// Identity for ASCII; an astral code point contributes its two surrogate
// units. Accepts the JNI modified-UTF-8 forms too (C0 80 for U+0000 and
// byte-encoded surrogates count as one unit each, as charCodeAt would see
// them); malformed bytes decode to U+FFFD.
std::vector<unsigned char> Utf16LowBytes(const std::string& utf8) {
  std::vector<unsigned char> units;
  units.reserve(utf8.size());
  size_t i = 0;
  while (i < utf8.size()) {
    const unsigned char first = static_cast<unsigned char>(utf8[i]);
    uint32_t codePoint = 0xfffd;
    size_t width = 1;
    if (first < 0x80) {
      codePoint = first;
    } else if (first >= 0xc2 && first <= 0xdf && i + 1 < utf8.size()) {
      const unsigned char b1 = static_cast<unsigned char>(utf8[i + 1]);
      if ((b1 & 0xc0) == 0x80) {
        codePoint = ((first & 0x1f) << 6) | (b1 & 0x3f);
        width = 2;
      }
    } else if (first == 0xc0 && i + 1 < utf8.size() &&
               static_cast<unsigned char>(utf8[i + 1]) == 0x80) {
      codePoint = 0;
      width = 2;
    } else if (first >= 0xe0 && first <= 0xef && i + 2 < utf8.size()) {
      const unsigned char b1 = static_cast<unsigned char>(utf8[i + 1]);
      const unsigned char b2 = static_cast<unsigned char>(utf8[i + 2]);
      if ((b1 & 0xc0) == 0x80 && (b2 & 0xc0) == 0x80 &&
          !(first == 0xe0 && b1 < 0xa0)) {
        codePoint = ((first & 0x0f) << 12) | ((b1 & 0x3f) << 6) | (b2 & 0x3f);
        width = 3;
      }
    } else if (first >= 0xf0 && first <= 0xf4 && i + 3 < utf8.size()) {
      const unsigned char b1 = static_cast<unsigned char>(utf8[i + 1]);
      const unsigned char b2 = static_cast<unsigned char>(utf8[i + 2]);
      const unsigned char b3 = static_cast<unsigned char>(utf8[i + 3]);
      if ((b1 & 0xc0) == 0x80 && (b2 & 0xc0) == 0x80 &&
          (b3 & 0xc0) == 0x80 && !(first == 0xf0 && b1 < 0x90) &&
          !(first == 0xf4 && b1 > 0x8f)) {
        codePoint = ((first & 0x07) << 18) | ((b1 & 0x3f) << 12) |
                    ((b2 & 0x3f) << 6) | (b3 & 0x3f);
        width = 4;
      }
    }
    i += width;
    if (codePoint <= 0xffff) {
      units.push_back(static_cast<unsigned char>(codePoint & 0xff));
    } else {
      codePoint -= 0x10000;
      units.push_back(static_cast<unsigned char>((0xd800 + (codePoint >> 10)) & 0xff));
      units.push_back(static_cast<unsigned char>((0xdc00 + (codePoint & 0x3ff)) & 0xff));
    }
  }
  return units;
}

}  // namespace

uint32_t Murmur3_32(const std::string& key, uint32_t seed) {
  // The TS reference emulates 32-bit multiplication with 16-bit halves;
  // native uint32_t arithmetic is exactly (x * y) mod 2^32, so the halved
  // dance collapses to plain multiplies. Input bytes are the low byte of each
  // UTF-16 code unit (charCodeAt & 0xff), not the UTF-8 bytes: hashing UTF-8
  // would bucket non-ASCII keys differently from the TS reference.
  const uint32_t c1 = 0xcc9e2d51;
  const uint32_t c2 = 0x1b873593;
  const std::vector<unsigned char> units = Utf16LowBytes(key);
  static const unsigned char kEmpty = 0;
  const unsigned char* data = units.empty() ? &kEmpty : units.data();
  const size_t len = units.size();
  const size_t nblocks = len / 4;
  uint32_t h1 = seed;

  for (size_t i = 0; i < nblocks; i++) {
    uint32_t k1 = static_cast<uint32_t>(data[i * 4]) |
                  (static_cast<uint32_t>(data[i * 4 + 1]) << 8) |
                  (static_cast<uint32_t>(data[i * 4 + 2]) << 16) |
                  (static_cast<uint32_t>(data[i * 4 + 3]) << 24);
    k1 *= c1;
    k1 = (k1 << 15) | (k1 >> 17);
    k1 *= c2;
    h1 ^= k1;
    h1 = (h1 << 13) | (h1 >> 19);
    h1 = h1 * 5 + 0xe6546b64;
  }

  uint32_t k1 = 0;
  const unsigned char* tail = data + nblocks * 4;
  switch (len & 3) {
    case 3:
      k1 ^= static_cast<uint32_t>(tail[2]) << 16;
      [[fallthrough]];
    case 2:
      k1 ^= static_cast<uint32_t>(tail[1]) << 8;
      [[fallthrough]];
    case 1:
      k1 ^= tail[0];
      k1 *= c1;
      k1 = (k1 << 15) | (k1 >> 17);
      k1 *= c2;
      h1 ^= k1;
  }

  h1 ^= static_cast<uint32_t>(len);
  h1 ^= h1 >> 16;
  h1 *= 0x85ebca6b;
  h1 ^= h1 >> 13;
  h1 *= 0xc2b2ae35;
  h1 ^= h1 >> 16;
  return h1;
}

bool IsInRollout(double rollout, const std::string& uuid) {
  return static_cast<double>(Murmur3_32(uuid) % 100) < rollout;
}

bool IsMirrorRetryableCode(const std::string& code) {
  return code != "PATCH_FAILED";
}

namespace {

// ^[a-z][a-z0-9+.-]*:\/\/ (case-insensitive)
bool HasExplicitScheme(const std::string& s) {
  size_t i = 0;
  auto isAlpha = [](char c) {
    return (c >= 'a' && c <= 'z') || (c >= 'A' && c <= 'Z');
  };
  if (i >= s.size() || !isAlpha(s[i])) {
    return false;
  }
  i++;
  while (i < s.size()) {
    char c = s[i];
    if (isAlpha(c) || (c >= '0' && c <= '9') || c == '+' || c == '.' ||
        c == '-') {
      i++;
    } else {
      break;
    }
  }
  return s.compare(i, 3, "://") == 0;
}

}  // namespace

Value JoinUrls(const Value& paths, const Value& fileName) {
  if (!fileName.Truthy()) {
    return Value::Undefined();
  }
  Value urls = Value::Array();
  for (const auto& path : paths.elements()) {
    std::string normalized = path.AsString();
    while (!normalized.empty() && normalized.back() == '/') {
      normalized.pop_back();
    }
    std::string base = HasExplicitScheme(normalized)
                           ? normalized
                           : "https://" + normalized;
    urls.Push(Value::String(base + "/" + fileName.AsString()));
  }
  return urls;
}

Value OrderEndpointCandidates(const Value& endpoints, double randomSample) {
  Value deduped = Value::Array();
  for (const auto& endpoint : endpoints.elements()) {
    if (!endpoint.Truthy()) {
      continue;
    }
    bool seen = false;
    for (const auto& kept : deduped.elements()) {
      // Set membership is ===: a stray non-string entry (a number, `true`)
      // dedupes against equal primitives only, never against strings.
      if (Value::StrictEquals(kept, endpoint)) {
        seen = true;
        break;
      }
    }
    if (!seen) {
      deduped.Push(endpoint);
    }
  }
  const size_t n = deduped.Size();
  if (n < 2) {
    return deduped;
  }
  const double idx = std::isfinite(randomSample)
                         ? std::floor(randomSample * static_cast<double>(n))
                         : 0;
  // Validate in floating-point space before converting to size_t: converting
  // NaN, infinity, or an out-of-range value is undefined behavior in C++.
  const size_t first = idx <= 0 ? 0
                                : (idx >= static_cast<double>(n - 1)
                                       ? n - 1
                                       : static_cast<size_t>(idx));
  Value ordered = Value::Array();
  ordered.Push(deduped.At(first));
  for (size_t i = 0; i < n; i++) {
    if (i != first) {
      ordered.Push(deduped.At(i));
    }
  }
  return ordered;
}

Value BuildCheckRequestBody(const Value& input) {
  Value body = Value::Object();
  // Caller extras go in FIRST: they may add fields but never override the
  // identity fields the server keys its decision on (mirrors the TS
  // `{ ...extra, packageVersion, hash, ... }` spread, including key order).
  const Value& extra = input.Get("extra");
  if (extra.IsObject()) {
    for (const auto& member : extra.members()) {
      body.Set(member.first, member.second);
    }
  }
  body.Set("packageVersion", input.Get("packageVersion"));
  body.Set("hash", input.Get("currentVersion"));
  body.Set("buildTime", input.Get("buildTime"));
  body.Set("cInfo", input.Get("cInfo"));
  const Value& diffV = input.Get("supportedDiffVersion");
  if (diffV.Truthy()) {
    body.Set("diffV", diffV);
  } else {
    body.Remove("diffV");
  }
  const Value& bundleHash = input.Get("bundleHash");
  if (bundleHash.Truthy()) {
    body.Set("bundleHash", bundleHash);
  } else {
    body.Remove("bundleHash");
  }
  if (input.Get("isDev").Truthy()) {
    body.Remove("buildTime");
  }
  // Transitional dual form (mirrors the TS side): nested copy of the caller
  // extras next to the legacy root-level spread.
  if (extra.IsObject() && !extra.members().empty()) {
    Value nested = Value::Object();
    for (const auto& member : extra.members()) {
      nested.Set(member.first, member.second);
    }
    body.Set("extra", std::move(nested));
  }
  return body;
}

Value ResolveCheckResult(const Value& rootInfo, const Value& identity) {
  Value rootResult = Value::Object();
  for (const auto& member : rootInfo.members()) {
    if (member.first != "expVersion") {
      rootResult.Set(member.first, member.second);
    }
  }
  const Value& expVersion = rootInfo.Get("expVersion");
  const Value& currentVersion = identity.Get("currentVersion");
  // expVersion?.config?.rollout?.[identity.packageVersion] — Get on a
  // non-object returns Undefined, mirroring optional chaining.
  const Value& rollout = expVersion.Get("config").Get("rollout").Get(
      identity.Get("packageVersion").AsString());
  if (rootResult.Get("update").Truthy() && expVersion.Truthy() &&
      rollout.IsNumber()) {
    if (IsInRollout(rollout.AsNumber(), identity.Get("uuid").AsString())) {
      const Value& expHash = expVersion.Get("hash");
      if (expHash.IsString() && !expHash.AsString().empty() &&
          Value::StrictEquals(expHash, currentVersion)) {
        Value upToDate = Value::Object();
        upToDate.Set("upToDate", Value::Bool(true));
        return upToDate;
      }
      Value info = Value::Object();
      info.Set("update", Value::Bool(true));
      for (const auto& member : expVersion.members()) {
        info.Set(member.first, member.second);
      }
      if (rootResult.Get("paths").Truthy()) {
        info.Set("paths", rootResult.Get("paths"));
      }
      return info;
    }
  }
  const Value& rootHash = rootResult.Get("hash");
  if (rootResult.Get("update").Truthy() && rootHash.IsString() &&
      !rootHash.AsString().empty() &&
      Value::StrictEquals(rootHash, currentVersion)) {
    Value upToDate = Value::Object();
    upToDate.Set("upToDate", Value::Bool(true));
    return upToDate;
  }
  return rootResult;
}

namespace {

Value DeclineDownload(const char* reason) {
  Value none = Value::Object();
  none.Set("action", Value::String("none"));
  none.Set("reason", Value::String(reason));
  return none;
}

}  // namespace

Value DecideDownload(const Value& info, const Value& identity, bool isDev) {
  const Value& hash = info.Get("hash");
  Value paths = info.Get("paths");
  if (paths.IsUndefined() || paths.kind() == Value::Kind::Null) {
    paths = Value::Array();  // info.paths ?? [] — a server `null` too
  }
  if (!info.Get("update").Truthy() || !hash.Truthy()) {
    return DeclineDownload("noUpdate");
  }
  if (Value::StrictEquals(hash, identity.Get("currentVersion"))) {
    return DeclineDownload("alreadyCurrent");
  }
  const Value& rolledBack = identity.Get("rolledBackVersion");
  if (rolledBack.Truthy() && Value::StrictEquals(hash, rolledBack)) {
    return DeclineDownload("rolledBack");
  }
  Value attempts = Value::Array();
  auto pushAttempt = [&attempts](const char* type, const Value& urls) {
    if (urls.IsArray() && urls.Size() > 0) {
      Value attempt = Value::Object();
      attempt.Set("type", Value::String(type));
      attempt.Set("urls", urls);
      attempts.Push(std::move(attempt));
    }
  };
  if (!isDev) {
    pushAttempt("diff", JoinUrls(paths, info.Get("diff")));
    pushAttempt("pdiff", JoinUrls(paths, info.Get("pdiff")));
  }
  Value fullUrls = JoinUrls(paths, info.Get("full"));
  pushAttempt("full", fullUrls);

  const bool devNoop =
      isDev && !(fullUrls.IsArray() && fullUrls.Size() > 0);
  if (attempts.Size() == 0 && !devNoop) {
    return DeclineDownload("noArtifact");
  }

  Value decision = Value::Object();
  decision.Set("action", Value::String("download"));
  decision.Set("hash", hash);
  decision.Set("attempts", std::move(attempts));
  decision.Set("devNoop", Value::Bool(devNoop));
  return decision;
}

bool ShouldActivateAfterDownload(const Value& info,
                                 const std::string& afterDownload) {
  return afterDownload == "setNeedUpdate" ||
         info.Get("config").Get("forceBoot").Truthy();
}

bool IsValidCheckResult(const Value& root) {
  if (!root.IsObject()) {
    return false;
  }
  const Value& upToDate = root.Get("upToDate");
  const Value& update = root.Get("update");
  const Value& expired = root.Get("expired");
  const Value& paused = root.Get("paused");
  return upToDate.IsBool() || update.IsBool() || expired.IsBool() ||
         paused.IsString();
}

bool IsValidCheckResponse(const std::string& responseText) {
  bool ok = false;
  Value root = flowjson::Parse(responseText, &ok);
  return ok && IsValidCheckResult(root);
}

Value HandleCheckResponse(const std::string& responseText,
                          const Value& identity, bool isDev,
                          const std::string& afterDownload) {
  bool ok = false;
  Value root = flowjson::Parse(responseText, &ok);
  if (!ok || !IsValidCheckResult(root)) {
    return DeclineDownload("invalidResponse");
  }
  Value resolved = ResolveCheckResult(root, identity);
  Value decision = DecideDownload(resolved, identity, isDev);
  if (decision.Get("action").AsString() == "download") {
    decision.Set("activate", Value::Bool(ShouldActivateAfterDownload(
                                 resolved, afterDownload)));
  }
  decision.Set("info", std::move(resolved));
  return decision;
}

}  // namespace updateflow
